### ISSUE TITLE
Basic CSS Cascade Issue with Client Components

### ISSUE THREAD

**Issue by tjdunklee** [vercel/next.js#68207]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/nextjs-css-order-issue-ltgl6n

### To Reproduce

1. Start the application in dev mode
2. Notice the background color of the header

### Current vs. Expected behavior

I would expect the background color of the header to be the light green but instead it is light red. This looks like it is being caused by the way CSS is being compiled. I would expect the global CSS to come before the CSS from the client component, but it is the opposite. This breaks the normal CSS cascade behavior.

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
  Available memory (MB): 4102
  Available CPU cores: 2
Binaries:
  Node: 20.11.1
  npm: 10.2.4
  Yarn: 1.22.19
  pnpm: 8.15.4
Relevant Packages:
  next: 14.2.5 // Latest available version is detected (14.2.5).
  eslint-config-next: 14.2.1
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local), next start (local)

### Additional context

This only happens if the `<Header>` is client component.

**Comment by tjdunklee** [comment]
It's possible this is a duplicate of #64921 but my situation seems much simpler. Hope it helps!

**Comment by AliRazaDev27** [comment]
Read the following article for detailed information.
https://nextjs.org/docs/app/building-your-application/styling/css-modules
First option is to avoid the naming conflict, if not possible then you have order your style imports in the way you want to apply them. look at the image,
![image](https://github.com/user-attachments/assets/5299fad1-b5a4-4358-94c7-d723a65dadd1)

**Comment by tjdunklee** [comment]
Thanks for the response @AliRazaDev27. Importing the global CSS in to the client component does fix the issue, but that seems like an unsustainable solution if you have a large amount of client components. It seems very odd that the cascade works just fine as a server component, but has an incorrect order as a client component. I would assume that any component nested within a layout would include its CSS after the global CSS imported into the layout.

**Comment by ksachin7** [comment]
Just change your header className and remove header from it. `styles.header` is scoped to the component, while combining `header` from a global class was causing the issue.

- Use `styles.header` if you want the light-green color.
- Use `header ${styles.header}`  when you want to combine styles from both global and module css. But then if you use same style eg. `background-color` in both global and module css it will use `background-color` from global styles.

While combining global and module styles can be useful, it should be done with caution to avoid styling conflicts and maintain the benefits of modular styling.

![Screenshot 2024-08-20 at 7 24 29 PM](https://github.com/user-attachments/assets/da14d2f8-cfab-4d4d-9461-f699fe0b61f1)
![Screenshot 2024-08-20 at 7 28 54 PM](https://github.com/user-attachments/assets/a8181920-270a-43ec-9a16-3af69fc1c976)

**Comment by tjdunklee** [comment]
Thanks for your response @ksachin7. Your solution definitely works, but it really is just a workaround for the root issue. I still think the order of the CSS files is incorrect which causes basic CSS cascade issues. You should be able to override a global CSS style using the cascade in a child component.

**Comment by wataru-chocola** [comment]
Does anyone know a workaround? This seems like a critical issue to me.

**Comment by wataru-chocola** [comment]
I've done a bit of investigation and found a problem that might be causing the issue, so I'd like to share my findings.
I'm not very familiar with webpack, so please forgive me if I'm off the mark.

## Possible cause

In SC, CSS imports and client component imports are processed by [`injectClientEntryAndSSRModules` in flight-client-entry-plugin](https://github.com/vercel/next.js/blob/576e04c12d8f7311e5dc1a1b21e743e5ba88ba8f/packages/next/src/build/webpack/plugins/flight-client-entry-plugin.ts#L739).
I suspect that the sorting logic within this function is not behaving as expected due to an issue (detailed below), which in turn results in the module array being passed to `next-flight-client-entry-loader` having an incorrect order, ultimately leading to this problem.

The sorting logic in `injectClientEntryAndSSRModules` is as follows:

```javascript
    const modules = Object.keys(clientImports)
      .sort((a, b) => (regexCSS.test(b) ? 1 : a.localeCompare(b)))
      .map((clientImportPath) => ({
        request: clientImportPath,
        ids: [...clientImports[clientImportPath]],
      }))
```

Please note that `regexCSS.test()` is called only on `b` in sort function.

The way the sort function (`sortFn`) is invoked is dependent on the JavaScript engine implementation, meaning that the order of calls is not guaranteed. Therefore, `sortFn` is expected to satisfy [certain properties such as anti-symmetry and transitivity](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort), but this implementation does not.

Due to this, the order of comparison between imports may vary based on how `sortFn` is invoked.

- `sortFn("a.css", "B.tsx")` : `["a.css", "B.tsx"]` (`regexCSS` failed, so use locale ordering)
- `sortFn("B.tsx", "a.css")` : `["a.css", "B.tsx"]` (`regexCSS` succeeded)- 
- `sortFn("b.css", "A.tsx")` : `["A.tsx", "b.css"]` (`regexCSS` failed, so use locale ordering)
- `sortFn("A.tsx", "b.css")` : `["b.css", "A.tsx"]` (`regexCSS` succeeded)

In other words, the order of styles will depend on the engine's sort implementation and the locale ordering of the imported paths.

## Possible fix

I don't fully understand the exact intention behind using regexCSS in this sort, but if the goal is to prioritize CSS imports over other imports, it should be written differently.

```javascript
    const modules = Object.keys(clientImports)
      .sort((a, b) => (regexCSS.test(b) ? 1 : regexCSS.test(a) ? -1 : a.localeCompare(b)))
      .map((clientImportPath) => ({
        request: clientImportPath,
        ids: [...clientImports[clientImportPath]],
      }))
```

**Comment by mysticflute** [comment]
I agree that this is a major CSS cascade issue. The code example here is simple so that it can repro what's going on, but trying to work around it isn't the answer.

In more complicated scenarios, this issue makes using client components very difficult. In my case I'm using normalize.css, which sets a bunch of default styles and resets. It must be first in the CSS cascade, allowing all subsequent styles to override it. Even though I'm including this first in my layout it's being ordered after client component CSS. So the normalize styles start overriding some of the client component styles.

**Comment by marcoshernanz** [comment]
Thanks for the clear reproduction. I confirmed this still reproduces with `next@16.4.0-canary.8` when using Webpack in both `next dev` and `next build`, while it does not reproduce with the default Turbopack path. This is the same Client Component CSS ordering bug already tracked in #65480, so I’m closing this as a duplicate.
